streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -122,7 +126,15 @@ public static FileLock lockStateDirectory(File stateDir) throws IOException {
 
     private static FileLock lockStateDirectory(File stateDir, int retry) throws IOException {
         File lockFile = new File(stateDir, ProcessorStateManager.LOCK_FILE_NAME);
-        FileChannel channel = new RandomAccessFile(lockFile, "rw").getChannel();
+        FileChannel channel = null;
+        synchronized (channels) {
+            channel = channels.get(lockFile);
+            if (channel == null) {
+                channel = new RandomAccessFile(lockFile, "rw").getChannel();
+                channels.put(lockFile, channel);
```

Comment:
Is the current implementation vulnerable to the following race condition?

1. thread1 grabs `CHANNELS` lock, get the channel object from the map; and then exists the `CHANNELS` lock;

2. thread2 grabs the lock, release the lock on it, close the channel, remove from the map.

3. thread1 now calls `channel.tryLock`, which throws `ClosedChannelException`.